pywebpush/tests/test_webpush.py

```diff
@@ -446,16 +446,15 @@ def test_exception(self):
         assert "{}".format(exp) == "WebPushException: foo"
         # Really should try to load the response to verify, but this mock
         # covers what we need.
-        response = Mock(spec=Response)
```

Comment:
Why are you breaking this out of the `Mock`? 

Ideally, this test should check that the error being returned contains the specified JSON blob. (That's the reason that I note that we should try to load the response to verify.) I'd like to keep this as a Mock so that we can do those checks later.

Reply:
the reason was that I wanted a test that will fail if object's `__bool__` method returns `False` (which was the case of `Response` object with status code 401 which is falsy) before I introduced changes to have certainty that everything works properly and the test passes

instance of `Mock` is truthy which caused to give a false impression that in case of 401 status code the error will also contain response details (and in the real world scenario the error did not if `Response` object passed to `WebPushException` contained 401 status code)

> Ideally, this test should check that the error being returned contains the specified JSON blob

I'm not sure if I got this, but using `Response`, such check is also performed in one of `assert` statements below (the only difference here is that we load JSON blob to `_content` attribute, which is a private attribute so this might be considered as bad practice to rely on private interface)

however, I guess that after introducing changes to check if `.response` attribute is not `None` the usage of `Mock` object is justified as we only check if `.response` object has something assigned 🤔 

I'll revert this back to use `Mock` object if you prefer this approach :smile:

Reply:
changed back to use `Mock` instead of `Response`

alternatively we could rely on `MagicMock` to properly mock `__bool__` behaviour, e.g.

```python
response = MagicMock(spec=Response)
response.text = (
    '{"code": 401, "errno": 109, "error": '
    '"Unauthorized", "more_info": "http://'
    "autopush.readthedocs.io/en/latest/htt"
    'p.html#error-codes", "message": "Requ'
    "est did not validate missing authoriz"
    'ation header"}'
)
response.json.return_value = json.loads(response.text)
response.status_code = 401
response.reason = "Unauthorized"
response.__bool__.return_value = False  # Response object is falsy in case of status code between 400 and 599
```

but this change is only to be 100% sure that in case of accidentally changing logic back to:

```python
...
    def __str__(self):
        extra = ""
        if self.response:
            try:
                extra = ", Response {}".format(
...
```
the test will indicate that this logic doesn't work properly

wdyt?

Reply:
Ah! That's a pretty good reason. Yeah, I'm also not a big fan of loading to a private interface since that can cause problems later (in the unlikely event that `requests` ever decides to change their internal structure. Hasn't happened for them, but there have been a bunch of others that have so I've got the scars for that.)

Getting back to the core idea of the bug, I absolutely agree that this check was far too loose and just checking truthiness is insufficient. Having it check for `None` is good. Let's land what you've got and I'll try to figure out a more rigorous test for this thing. 

Thank You for giving me this to think about! (This is what I get for being lazy when I was putting this all together. ;) )

Reply:
No problem, I'm more than happy, that I could help in some way 😄 

(To be honest, I've found out about this truthy/falsy behaviour of `requests.Response` based on status code, just before I've created a PR while I was playing around with `pywebpush` library 😅 )